CRM/MembershipExtras/Hook/Alter/CalculatedMembershipStatus.php

```diff
@@ -0,0 +1,336 @@
+      ]);
+      self::$memberShipStatuses = $membershipStatuses['values'];
+    }
+
+    if (count(self::$contributionStatusValueMap) == 0) {
+      $contributionStatuses = civicrm_api3('OptionValue', 'get', array(
+        'option_group_id' => "contribution_status",
+      ));
+
+      foreach ($contributionStatuses['values'] as $currentStatus) {
+        self::$contributionStatusValueMap[$currentStatus['name']] = $currentStatus['value'];
+      }
+    }
+  }
+
+  /**
+   * This hook is called when membership status is being calculated.
+   *
+   * @param array $calculatedStatus
+   *   Membership status details as determined by CiviCRM core
+   * @param array $arguments
+   *   Arguments passed in to calculate date
+   * @param array $membership
+   *   Membership details from the calling function
+   */
+  public function alterMembershipStatus(&$calculatedStatus, $arguments, $membership) {
+    $this->membership = $membership;
+    $this->calculationArguments = $arguments;
+
+    foreach (self::$memberShipStatuses as $currentStatus) {
+      $startEventIsArrearsRelated = stripos($currentStatus['start_event'], 'arrears') !== false;
+      $endEventIsArrearsRelated = stripos($currentStatus['end_event'], 'arrears') !== false;
+
+      if (!$startEventIsArrearsRelated && !$endEventIsArrearsRelated) {
+        // If we reach calculated status, we don't need to consider other options by weight.
+        if ($calculatedStatus['id'] == $currentStatus['id']) {
+          break;
+        }
+
+        // No arrears, so we continue with next status.
+        continue;
+      }
+
+      $startEvent = $this->checkEvent(
+        CRM_Utils_Array::value('start_event', $currentStatus),
+        $arguments['status_date'],
+        CRM_Utils_Array::value('start_event_adjust_unit', $currentStatus),
+        CRM_Utils_Array::value('start_event_adjust_interval', $currentStatus)
+      );
+
+      $endEvent = $this->checkEvent(
+        CRM_Utils_Array::value('end_event', $currentStatus),
+        $arguments['status_date'],
+        CRM_Utils_Array::value('start_event_adjust_unit', $currentStatus),
+        CRM_Utils_Array::value('start_event_adjust_interval', $currentStatus)
+      );
+
+      if ($startEvent && !$endEvent) {
+        $calculatedStatus['id'] = $currentStatus['id'];
+        $calculatedStatus['name'] = $currentStatus['name'];
+        break;
+      }
+    }
+  }
+
+  /**
+   * Checks if the given event has presented itself on the given membership.
+   *
+   * @param $event
+   *   Event to be checked
+   * @param string $referenceDate
+   *   Date to use as reference to check start date
+   * @param string $adjustUnit
+   *   Unit of time interval to be added to start date
+   * @param int $adjustInterval
+   *   Amount of time to be added to start date
+   *
+   * @return boolean
+   *   True if the event has occurred, false otherwise
+   */
+  private function checkEvent($event, $referenceDate, $adjustUnit, $adjustInterval) {

```

Comment:
due to this check https://github.com/compucorp/uk.co.compucorp.membershipextras/blob/824cd2cf95b42edeb2a3c4ac7f19f73035d4c641/CRM/MembershipExtras/Hook/Alter/CalculatedMembershipStatus.php#L52

the status to be passed to checkEvent method can be either in_arrears or not_arrears, so I don't see a reason to add a case to other types, and based on that we can remove this method and call checkInArrearsEvent and checkNotInArrearsEvent methods directly.

Reply:
I checked with Guanhuan and he said it was ok to mix arrears events with other events, so for example, I could create a status called 'Urgent Collection', starting one month before 'End date' and finishing in 'Not in arrears', so memberships with less than one month to go, with outstanding invoices for a payment plan are marked.